src/row_codec.rs

```diff
@@ -1,182 +1,230 @@
+///  |       u16      |      u16       |  var        | u64     | u8        | i64       | i64       | u32       |  var      |
+///  |----------------|----------------|-------------|---------|-----------|-----------|-----------|-----------|-----------|
+///  | key_prefix_len | key_suffix_len |  key_suffix | seq     | flags     | expire_ts | create_ts | value_len | value     |
+///  |---------------------------------------------------------------------------------------------------------------------|
 /// ```
 ///
-/// And for tombstones (with `meta[row_flags] & 0x01 == 1`):
+/// And for tombstones (flags & Tombstone == 1):
+///
 ///  ```txt
-///  |------------------------------------------------------|
-///  |       u16      |      u16       |  var        | var  |
-///  |----------------|----------------|-------------|------|
-///  | key_prefix_len | key_suffix_len |  key_suffix | meta |
-///  |------------------------------------------------------|
+///  |----------------------------------------------------------|-----------|
+///  |       u16      |      u16       |  var        | u64      | u8        |
+///  |----------------|----------------|-------------|----------|-----------|
+///  | key_prefix_len | key_suffix_len |  key_suffix | seq      | flags     |
+///  |----------------------------------------------------------------------|
 ///  ```
 ///
-/// | Field            | Type | Description                                   |
-/// |------------------|------|-----------------------------------------------|
-/// | `key_prefix_len` | `u16` | Length of the key prefix                     |
-/// | `key_suffix_len` | `u16` | Length of the key suffix                     |
-/// | `key_suffix`     | `var` | Suffix of the key                            |
-/// | `meta`           | `var` | Metadata                                     |
-/// | `value_len`      | `u32` | Length of the value                          |
-/// | `value`          | `var` | Value bytes                                  |
-pub(crate) fn encode_row_v0(
-    data: &mut Vec<u8>,
-    key_prefix_len: usize,
-    key_suffix: &[u8],
-    value: Option<&[u8]>,
-    row_features: &[RowFeature],
-    timestamp: Option<i64>,
-    expire_at: Option<i64>,
-) {
-    data.put_u16(key_prefix_len as u16);
-    data.put_u16(key_suffix.len() as u16);
-    data.put(key_suffix);
-
-    data.put(encode_meta(
-        row_features,
-        value.is_none(),
-        timestamp,
-        expire_at,
-    ));
-
-    if let Some(value) = value {
-        data.put_u32(value.len() as u32);
-        data.put(value);
-    }
+/// | Field            | Type  | Description                                            |
+/// |------------------|-------|--------------------------------------------------------|
+/// | `key_prefix_len` | `u16` | Length of the key prefix                               |
+/// | `key_suffix_len` | `u16` | Length of the key suffix                               |
```

Comment:
I am still reading the code. What is `key_suffix` used for?

Reply:
i remember that `key_suffix` was introduced in this issue: https://github.com/slatedb/slatedb/issues/237

it introduced a prefix encoding, which allows SST to become smaller when we have a common prefix of keys inside it.